services/users/session.go

```diff
@@ -98,3 +110,134 @@ func (us *UserService) ClearAllUsersSessionCache() {
 		us.cluster.SendClusterMessage(msg)
 	}
 }
+
+func (us *UserService) GetSessionByID(sessionID string) (*model.Session, error) {
+	return us.sessionStore.Get(context.Background(), sessionID)
+}
+
+func (us *UserService) RevokeSessionsFromAllUsers() error {
+	// revoke tokens before sessions so they can't be used to relogin
+	nErr := us.oAuthStore.RemoveAllAccessData()
+	if nErr != nil {
+		return DeleteAllAccessDataError
+	}
+	err := us.sessionStore.RemoveAllSessions()
+	if err != nil {
+		return err
+	}
+
+	us.ClearAllUsersSessionCache()
+	return nil
+}
+
+func (us *UserService) RevokeSessionsForDeviceId(userID string, deviceID string, currentSessionId string) error {
+	sessions, err := us.sessionStore.GetSessions(userID)
+	if err != nil {
+		return err
+	}
+	for _, session := range sessions {
+		if session.DeviceId == deviceID && session.Id != currentSessionId {
+			mlog.Debug("Revoking sessionId for userId. Re-login with the same device Id", mlog.String("session_id", session.Id), mlog.String("user_id", userID))
+			if err := us.RevokeSession(session); err != nil {
+				mlog.Warn("Could not revoke session for device", mlog.String("device_id", deviceID), mlog.Err(err))
+			}
+		}
+	}
+
+	return nil
+}
+
+func (us *UserService) RevokeSession(session *model.Session) error {
+	if session.IsOAuth {
+		if err := us.RevokeAccessToken(session.Token); err != nil {
+			return err
+		}
+	} else {
+		if err := us.sessionStore.Remove(session.Id); err != nil {
+			return DeleteSessionError
```

Comment:
Should we wrap to surface the underlying store error?

Reply:
Maybe you missed this :)

Reply:
lol, yeah 🤦